File: 0040-combination-sum-ii/0040-combination-sum-ii.go
import (
    "sort"
)

func combinationSum2(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    combos := make([][]int, 0)
    
    var dfs func([]int, int, []int)
    dfs = func(nums []int, targ int, path []int) {
        if targ < 0 {
            return
        }
        if targ == 0 {
            combos = append(combos, path)
            return
        }
        
        for i, n := range nums {
            if i > 0 && nums[i] == nums[i-1] {
                continue
            }
            newPath := make([]int, len(path))
            copy(newPath, path)
            dfs(nums[i+1:], targ-n, append(newPath, n))
        }
        return
    }
    
    dfs(candidates, target, []int{})
    return combos
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        self.combos = []
        candidates.sort()
        self.combo_sum_rec(candidates, target, [])
        return self.combos

    def combo_sum_rec(self, candidates: List[int], target: int, path: List[int]):
        if target == 0:
            self.combos.append(path)
            return
        
        if target < 0:
            return 
        
        for i in range(len(candidates)):
            if i > 0 and candidates[i] == candidates[i-1]:
                continue
            self.combo_sum_rec(candidates[i+1:], target-candidates[i], path + [candidates[i]])    
        return